### PR TITLE
Convert CommonJS modules to ES6, restructure code

### LINKED ISSUE #5: Mail --> Email
Look at the column header, this discrepancy also appears in mongoDB: `dpdmongo`

---

image

### PR DESCRIPTION
**Pull request by blms**
### Changes

<details>

<summary><b>Click to expand</b></summary>
<br />
Related to bundle refactor:
<br />
<br />

* Refactor all as ECMAScript (ES6), including backend code
  - All `require` --> `import`, `module.exports` --> `export`
  - Proper named imports in `Study.react.js` and `Matrix.d3.js`
* Move all backend code into `/server` directory for one-shot transpilation with babel
  - Update all path references to match new dir structure
* Use babel `(@babel/cli`) to transpile backend code
  - Store output in `/dist` (must be mounted on container)
  - Add script `npm run transpile` that only builds backend code
  - Modify webpack config accordingly
  - Add `build.sh` script to build with env vars
  - Start `pm2` with transpiled output (`/dist`)
  - Update `setup.sh` and `push.sh` scripts to use transpiled output
  - Update Singularity commands to mount `${state}/dpdash/dist` to `/sw/apps/dpdash/dist`
* Move config.js into `/server/configs` because ES6 imports cannot use  conditionals or variables (i.e. cannot be defined at runtime)
  - mv `${state}/dpdash/configs/dpdash.js` --> `${state}/dpdash/configs/dashboard/config.js`
  - Refactor `defaultStudyConfig.js` and `defaultUserConfig.js` as ES6 modules, update `configure.py` to generate `config.js` in ES6
  - Update Singularity commands to mount `${state}/dpdash/configs/dashboard` to `/sw/apps/dpdash/server/configs`
  - Remove references to unused environment variables
* Check if in browser when calling `window` in `stores/reducers.js`
* Update `README-PNL.md` to reflect change
* Prefer const and let over var (WIP)

Other:
* 🧹 Enforce consistent spacing and indentation
* Bump some dependencies
* Use `amqp://` instead of `amqps://` to avoid connection errors (it seems rabbit is running on `amqp://`)
* Reduce sleep time in `run.sh`
* Remove duplicate `defaultStudyConfig.js` and `defaultUserConfig.js`

</details>

**Breaking changes summary:**
* `/sw/apps/dpdash/dist` and `/sw/apps/dpdash/server/configs` must be mounted on the Singularity container, unless the user is mounting the entire `/sw/apps/dpdash` (e.g. as a dev instance) in which case subfolders `dist` and `server/configs` must be created.
  - The directory mounted as `/sw/apps/dpdash/server/configs` must contain `config.js`, `defaultStudyConfig.js`, and `defaultUserConfig.js`
  - The directory mounted as `/sw/apps/dpdash/dist` may be empty on creation, but must be populated using `singularity/run.sh` (via `npm run transpile`) 
* `${state}/dpdash/configs/dpdash.js` has been renamed to `${state}/dpdash/configs/dashboard/config.js`.

### Notes

It seems GitHub sees a lot of these moved files as "deleted" and "new", but only in the PR "files changed" tab and not in the individual commits. To see the actual changes to each file, please look at the individual commits. The big one is [here](https://github.com/PREDICT-DPACC/dpdash/pull/19/commits/88168cabe37f813d5134e56410dc75c5e8d9bf23).

### Testing

- [x] upon every update, build the Singularity image from scratch
  - `/data/predict/dpdash-ben-0507.sif`
- [x] delete `${state}` directory
- [x] run `init.sh` which will regenerate `${state}`
- [x] `import.py` into mongodb
- [x] check the website for expected results, browse various URIs

### CODE REVIEW

**Comment by blms** [comment]
It seems GitHub sees a lot of these moved files as "deleted" and "new", but only in the PR "files changed" tab and not in the individual commits. To see the actual changes to each file, please look at the individual commits. The big one is 88168ca.

**Comment by tashrifbillah** [comment]
Can you post the current structure?

> tree ${state}/dpdash/configs/

This is what it used to be so far:

```
├── dashboard
│   ├── config.schema
│   ├── defaultStudyConfig.js
│   └── defaultUserConfig.js
├── dpdash.js
├── dppy.conf
├── mongodb.conf
├── rabbitmq.config
└── supervisord.conf
```

**Comment by blms** [comment]
```
├── dashboard
│   ├── config.js
│   ├── config.schema
│   ├── defaultStudyConfig.js
│   └── defaultUserConfig.js
├── dppy.conf
├── mongodb.conf
├── rabbitmq.conf
├── rabbitmq-env.conf
└── supervisord.conf
```

**Comment by tashrifbillah** [comment]
Okay, what about `tree ${state}/ -P "*js"`? I am trying to get a better picture of where the new configs went.

**Comment by blms** [comment]
That's the only config movement that happened, moving `dpdash.js` to `dashboard/config.js`.

But here's that output. It also contains the `dist` directory tree, the result of running `npm run transpile` on the Singularity image while that directory is mounted.

<details>

<summary> Result of
<pre> tree ${state}/ -P "*js" </pre>
</summary>

```
├── dpdash
│   ├── celery
│   ├── configs
│   │   └── dashboard
│   │       ├── config.js
│   │       ├── defaultStudyConfig.js
│   │       └── defaultUserConfig.js
│   ├── dist
│   │   ├── app.js
│   │   ├── bin
│   │   │   └── www.js
│   │   ├── configs
│   │   │   ├── config.js
│   │   │   ├── defaultStudyConfig.js
│   │   │   └── defaultUserConfig.js
│   │   ├── routes
│   │   │   ├── index.js
│   │   │   └── users.js
│   │   ├── templates
│   │   │   ├── Account.template.js
│   │   │   ├── Admin.template.js
│   │   │   ├── Config.template.js
│   │   │   ├── DeepDive.template.js
│   │   │   ├── EditConfig.template.js
│   │   │   ├── Graph.template.js
│   │   │   ├── Login.template.js
│   │   │   ├── Main.template.js
│   │   │   ├── Register.template.js
│   │   │   ├── Resetpw.template.js
│   │   │   ├── Study.template.js
│   │   │   └── Subject.template.js
│   │   └── utils
│   │       ├── browserUtil.js
│   │       ├── crypto
│   │       │   └── hash.js
│   │       ├── dateConverter.js
│   │       ├── importer.js
│   │       ├── invertColor.js
│   │       ├── mongoUtil.js
│   │       └── passport
│   │           ├── ldap.js
│   │           ├── local-login.js
│   │           └── local-signup.js
│   ├── mongodb
│   │   ├── dbs
│   │   │   ├── diagnostic.data
│   │   │   └── journal
│   │   └── logs
│   ├── rabbitmq
│   │   ├── rabbit2
│   │   │   ├── msg_store_persistent
│   │   │   └── msg_store_transient
│   │   └── rabbit2-plugins-expand
│   ├── supervisord
│   │   └── logs
│   └── uploads
├── scripts
└── ssl
    ├── ca
    │   ├── certs
    │   └── private
    ├── client
    └── server
    
32 directories, 31 files
```

</details>

**Comment by tashrifbillah** [comment]
Where is this then?

> /sw/apps/dpdash/server/configs must contain config.js, defaultStudyConfig.js, and defaultUserConfig.js

I don't see any `server` directory.

**Comment by blms** [comment]
That directory is mounted from`${state}/dpdash/configs/dashboard`, visible in the tree above. You can see it contains those three files.

**Comment by blms** [comment]
> I don't see any `server` directory.

That's part of the breaking changes, it's now required that you mount the `server/configs` directory. You can see this in the updated docs:
https://github.com/PREDICT-DPACC/dpdash/blob/7865b9dffda513b2595d094c3d1957a90ae6e9ef/README-PNL.md#run


```
...
-B ${state}/dpdash/configs/dashboard:/sw/apps/dpdash/server/configs \
...
```

**Comment by blms** [comment]
`server` only exists inside of `/sw/apps/dpdash`, which only exists on the Singularity image. But we need `configs` to be a writeable subfolder of `server`, so we mount it from `${state}/dpdash/configs/dashboard`. Similarly, we need `/sw/apps/dpdash/dist` to be a writeable subfolder of `/sw/apps/dpdash`, so we mount it from `${state}/dpdash/dist`. (It might help to know that `dist`'s contents are the result of transpiling all of `server`, so its directory structure is identical.)

Hopefully this helps clear it up a little.

**Comment by tashrifbillah** [comment]
The PR is huge, more comments are coming later today ...

**Comment by blms** [comment]
> The PR is huge, more comments are coming later today ...

Yup, no rush!

File: README-PNL.md [hunk 1]
```diff
@@ -111,18 +112,31 @@ You can read the [Cookies](#cookies) section for details.
 
 Launch a DPdash instance as follows:
     
-    singularity run -B ${state}:/data -B ${data}:/project_data ${DPDASH_IMG} /sw/apps/dpdash/singularity/run.sh
+    singularity run \
+    -B ${state}:/data \
+    -B ${state}/dpdash/configs/dashboard:/sw/apps/dpdash/server/configs \
```
Comment (tashrifbillah) [inline README-PNL.md:117]: For making `require`-->`import ../` type changes, if the two-step mounting is essential, you probably don't have a choice but I am nervous about asking a user to mount in two steps as opposed to just one. 

By the way, did you already include the new mounts in the convenience scripts? The PR is huge, I didn't get there yet.

Comment (blms) [reply README-PNL.md:117]: I did! 

I do agree these mount statements are getting unwieldy. The mounts are still all in one terminal command but I wanted to make it more readable by splitting up the lines in the docs with `\ [newline]`. Hopefully users can handle it…

We could switch things around a little as I had started to do on a previous PR and move all of those `singularity` statements primarily into the scripts, and just say "advanced users may mount their own directories with the following command". I did notice while making these changes that some scripts (`init.sh` and `import.sh`) already contained a `singularity` command with hard coded mount statements, before I started working on DPdash, so it wasn't a completely new practice.

Comment (tashrifbillah) [reply README-PNL.md:117]: > The mounts are still all in one terminal command but I wanted to make it more readable by splitting up the lines in the docs with \ [newline]

I know, that's not the concern, this is:

> I do agree these mount statements are getting unwieldy.

Comment (blms) [reply README-PNL.md:117]: In that case, should we make the convenience scripts the preferred way of accessing these commands and make the manual bindings for advanced users? The mounts are required for the `import` statements.

Comment (tashrifbillah) [reply README-PNL.md:117]: Remind me--what's the challenge inherent with putting the contents of two folders into one and then mounting just one?

Comment (blms) [reply README-PNL.md:117]: I can't think of a way for these two folders to be mounted in the same bind statement / from the same folder, because `configs` is pre-transpilation and `dist` is post-transpilation. The flow is like this:

1. Configs are placed in `server/configs`
2. `npm run transpile` transpiles ALL contents of `/server`, including `/configs`, and dumps the result into `/dist`
3. Code in `/dist` is run by Node

In other words, we need to mount a directory (`configs`) that only lives inside of `server` and a directory (`dist`) that only lives outside of `server`, i.e. at the project root. If we mount both of them in one bind action, they'll either both or neither live inside of `server`.

Comment (tashrifbillah) [reply README-PNL.md:117]: Can you rephrase transpile?

Comment (blms) [reply README-PNL.md:117]: After writing it out and thinking further, I wonder if this is where that could happen:
> `npm run transpile` transpiles ALL contents of `/server`, including `/configs`, and dumps the result into `/dist`

Maybe I can configure babel to ignore `/dist` and make it a subdirectory of `/server`. Then both folders get mounted into `/server/local`, like `/server/local/dist` and `/server/local/configs`, or something like that.

It would only reduce the number of bind statements by one, but if you think that's worth it I can go ahead.

Comment (blms) [reply README-PNL.md:117]: > Can you rephrase transpile?

It's the process of automatically converting ES6 code into CommonJS or ES5 that Node and the browser can interpret, while retaining ES6's optimizations.

See [babel](https://babeljs.io/docs/en/) for more info.

Comment (tashrifbillah) [reply README-PNL.md:117]: > but if you think that's worth it

Definitely, minimize complexity at all costs in software designing. I shall rethink your strategy and give it a go later.

Comment (blms) [reply README-PNL.md:117]: > I shall rethink your strategy and give it a go later. 

Did you mean I should go ahead and make those changes or wait for your further comment?

Comment (tashrifbillah) [reply README-PNL.md:117]: > wait for your further comment?

Yes

File: singularity/configure.py [hunk 2]
```diff
@@ -293,12 +293,12 @@ def configure_supervisord(supervisor_id, config_path):
     return export_file(os.path.join(config_path, 'supervisord.conf'), configuration)
 
 def configure_dpdash(ca_path, key_path, cert_path, dpdash_port, mongo_port, rabbit_port, mongo_password, rabbit_password, session_secret, config_path, data_repo, dpdash_path, rabbit_host, mongo_host, secret):
-    configuration = '''var fs = require("fs");
-var cert = fs.readFileSync("%(cert_path)s");
-var key = fs.readFileSync("%(key_path)s");
-var ca = [fs.readFileSync("%(ca_path)s")];
+    configuration = '''import { readFileSync } from "fs";
```
Comment (tashrifbillah) [inline singularity/configure.py:296]: Did I get it right that this configuration is written to `dashboard/config.js`?

File: singularity/run.sh [hunk 3]
```diff
@@ -24,17 +24,14 @@ export dppy_config=/data/dpdash/configs/dppy.conf
...
-sleep 100
+# sleep 30 seconds
+sleep 30
 
 # vars for DPdash node app
-export DPDASH_CONFIG=/data/dpdash/configs/dpdash.js
 export DPDASH_UPLOADS=/data/dpdash/uploads
 export DPDASH_UPLOADS_CONFIG_SCHEMA=/data/dpdash/configs/dashboard/config.schema
-export DPDASH_DASHBOARD_CONFIG_DEFAULT=/data/dpdash/configs/dashboard/defaultUserConfig.js
-export DPDASH_DASHBOARD_CONFIG_DEFAULT_STUDY=/data/dpdash/configs/dashboard/defaultStudyConfig.js
```
Comment (tashrifbillah) [inline singularity/run.sh:35]: Are the definitions no longer required because of mounting in the new way?

File: singularity/start.sh [hunk 4]
```diff
@@ -23,7 +28,35 @@ source ./varcheck.sh
 
 if [ $shellmode = true ]
 then
-    singularity shell -B ${state}:/data -B ${data}:/project_data ${DPDASH_IMG}
+    if [ $devmode = true ]
```
Comment (tashrifbillah) [inline singularity/start.sh:37]: The `devmode` poses a dual requirement and should be omitted. We should just check:

> if [ -z ${DPDASH_DEV_DIR} ] || [ ! -d "$DPDASH_DEV_DIR" ]

File: views/Main.render.react.js [hunk 5]
```diff
@@ -1,10 +1,10 @@
...
+import { render } from 'react-dom';
 import { Provider } from 'react-redux';
 
 import MainPage from './Main.react';
 import store from '../stores/store';
 
 const reduxStore = store();
 
-render(<Provider store={reduxStore}><MainPage/></Provider>, document.getElementById('main'));
+render(<Provider store={reduxStore}><MainPage /></Provider>, document.getElementById('main'));
```
Comment (tashrifbillah) [inline views/Main.render.react.js:10]: Just a question, did you do this white space adjustment through https://codebeautify.org/ like tool or manually?

File: singularity/configure.py (hunk 2 above)
Comment (blms) [reply singularity/configure.py:296]: Correct.

File: singularity/run.sh (hunk 3 above)
Comment (blms) [reply singularity/run.sh:35]: Yes.

File: README-PNL.md (hunk 1 above)
Comment (tashrifbillah) [reply README-PNL.md:117]: No need to do any modification on this, thank you.

File: singularity/start.sh (hunk 4 above)
Comment (blms) [reply singularity/start.sh:37]: By "dual requirement", do you mean that one is required to both set `DPDASH_DEV_DIR` as well as use the `-d` flag, triggering the `devmode=true` statement?

The reason I did this is because I have `DPDASH_DEV_MODE` defined in my `.env`. I would like to keep it there and still be able to choose whether or not I want to mount it by choosing whether or not to use `-d`. If we only check that it is defined and is a directory, then it will always enter "dev mode" (i.e. it will always be mounted in this script). 

hat was the purpose of the `-d` flag—sometimes I don't want to mount that directory but just use the read-only `/sw/apps/dpdash` that's on the image, and in those cases, I can just omit the `-d` flag.

Or did you mean something else by "dual requirement"?

Comment (tashrifbillah) [reply singularity/start.sh:37]: Understood, forget about the modification!

Comment (blms) [reply singularity/start.sh:37]: If you mean that it requires both the `-s` and the `-d` flag for that code to execute at all, I could move this block that checks if `devmode` is `true` outside of the "`shellmode` is `true`" check, so that you could just `singularity exec ... run.sh` while mounting the dev directory by using the `-d` flag without `-s`. It was just a convenience script for me, and I was always running Singularity in shell mode when I mounted the dev directory, but it makes sense to open that up more.

File: views/Main.render.react.js (hunk 5 above)
Comment (blms) [reply views/Main.render.react.js:10]: Through a tool, I used the [Prettier VSCode extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode).

File: singularity/start.sh (hunk 4 above)
Comment (blms) [reply singularity/start.sh:37]: > I could move this block that checks if `devmode` is true outside of the "`shellmode` is true"

Even if that's not what you meant, I think I'll make that change anyway!

**Review by tashrifbillah — APPROVED** [review]
Very good!

**Comment by blms** [comment]
Closes #5.
